Add type annotations to the PauliTerm class

rht opened the pull request:
Description
-----------
Only PauliTerm is annotated to keep this PR small.

Checklist
---------

- [x] The above description motivates these changes.
- [x] There is a unit test that covers these changes.
- [x] All new and existing tests pass locally and on Semaphore.
- [x] Parameters have type hints with [PEP 484 syntax](https://www.python.org/dev/peps/pep-0484/).
- [x] Functions and classes have useful sphinx-style docstrings.
- [x] (New Feature) The docs have been updated accordingly.
- [x] (Bugfix) The associated issue is referenced above using
      [auto-close keywords](https://help.github.com/en/articles/closing-issues-using-keywords).
- [x] The [changelog](https://github.com/rigetti/pyquil/blob/master/CHANGELOG.md) is updated,
      including author and PR number (@username, gh-xxx).

rht commented:
Based on #147.

appleby commented on pyquil/paulis.py:
```diff
@@ -70,7 +70,7 @@ class PauliTerm(object):
     """
 
-    def __init__(self, op: str, index: int, coefficient: Union[int, float, complex] = 1.0):
+    def __init__(self, op: str, index: int, coefficient: Union[int, float, complex] = 1.0) -> None:
```
I think you can drop the `-> None` return type annotation here.

rht replied:
OK, it was my habit from Zulip's annotation (e.g. https://github.com/zulip/zulip/blob/master/zerver/forms.py#L92). I suppose this convention varies across projects. Will remove the `-> None`.

rht commented:
I have added a description of the PR into the changelog.

rht commented:
> Also, does mypy report any errors/warnings when running with --strict ? E.g.

There are lots of errors/warnings still because not everything inside paulis.py has been annotated yet. Hopefully the number of errors/warnings become tractable as more of the functions become annotated.

rht commented on pyquil/paulis.py:
```diff
@@ -203,7 +203,7 @@ def _multiply_factor(self, factor, index):
         return new_term
 
-    def __mul__(self, term):
+    def __mul__(self, term: Union['PauliTerm', 'PauliSum', 'Number']) -> 'PauliTerm':
```
I am basing the return type annotation on the docstring. But this should in general be a `Union['PauliTerm', 'PauliSum']`, because sometimes it does return a `PauliSum`.

appleby replied:
In that case, I would vote update the docstring and have the type hint reflect the actual return type.

Or better yet, just remove all these `:rtype:` lines from the docstrings for any methods that have a return type hint. We use a sphinx plugin to automatically add the return type if a type hint is present, so they are now redundant.

rht replied:
Updated the return type. For removing `:rtype:`, I think it should be done in a separate PR, to keep things modular.

appleby replied:
It's not a blocker, but unless you're planning to file an immediate follow-on PR, my preference would be to handle it in this PR. Otherwise, these things tend to slip through the cracks and eventually the docstring types become stale or out-of-sync with the type hints.

appleby commented on pyquil/paulis.py:
```diff
@@ -203,13 +203,13 @@ def _multiply_factor(self, factor, index):
         return new_term
 
-    def __mul__(self, term):
+    def __mul__(self, term: Union['PauliTerm', 'PauliSum', 'Number']) -> Union['PauliTerm', 'PauliSum']:
```
```suggestion
    def __mul__(self, term: Union['PauliTerm', 'PauliSum', Number]) -> Union['PauliTerm', 'PauliSum']:
```

appleby commented on pyquil/paulis.py:
```diff
@@ -318,7 +318,7 @@ def compact_str(self):
 
     @classmethod
-    def from_list(cls, terms_list, coefficient=1.0):
+    def from_list(cls, terms_list, coefficient: float = 1.0) -> 'PauliTerm':
```
```suggestion
    def from_list(cls, terms_list: List[Tuple(str, int)], coefficient: float = 1.0) -> 'PauliTerm':
```

Actually, `Iterable` would be preferable to `List`, but since the method is called `from_list`, go for consistency I guess.

appleby commented on pyquil/paulis.py:
```diff
@@ -389,7 +389,7 @@ def from_compact_str(cls, str_pauli_term):
         return op
 
-    def pauli_string(self, qubits=None):
+    def pauli_string(self, qubits=None) -> str:
```
```suggestion
    def pauli_string(self, qubits: Optional[Iterable[int]] = None) -> str:
```

Might have to import `Iterable`. Also, update the docstring to say `Iterable` instead of `list` where appropriate.

appleby commented on pyquil/paulis.py:
```diff
@@ -421,21 +421,21 @@ def pauli_string(self, qubits=None):
 
 
-def sI(q=None):
+def sI(q: int = None) -> PauliTerm:
```
```suggestion
def sI(q: Optional[int] = None) -> PauliTerm:
```

rht commented:
@appleby I have applied your patches to the PR. I will add annotations to the remaining 4 methods, and I will report when I'm done.

rht commented:
@appleby I added `operations_as_set`, `__hash__`, `program`, `__iter__` annotations in the second commit.

appleby commented on pyquil/paulis.py:
```diff
@@ -170,22 +170,22 @@ def copy(self):
         return self._ops.get(i, "I")
 
-    def __iter__(self):
+    def __iter__(self) -> Iterable[Tuple[int, str]]:
```
```suggestion
    def __iter__(self) -> Iterator[Tuple[int, str]]:
```

mypy complains about `Iterable` not having a `__next__` attr. This actually [returns an `Iterator`](https://docs.python.org/3/library/stdtypes.html#generator-types).

rht replied:
Fixed.

appleby commented on pyquil/paulis.py:
```diff
@@ -203,13 +203,13 @@ def _multiply_factor(self, factor, index):
         return new_term
 
-    def __mul__(self, term):
+    def __mul__(self, term: Union['PauliTerm', 'PauliSum', Number]) -> Union['PauliTerm', 'PauliSum']:
```
Ultimately, we'll probably want to convert this to use [function overloading](https://mypy.readthedocs.io/en/latest/more_types.html#function-overloading) to help mypy work out the relationship between `term`'s type and the return type, since `Union` return types are [sometimes cumbersome](https://github.com/python/mypy/issues/1693).

But I think it's fine to leave it as is for now, and address that as this module gets closer to being fully typed

appleby commented on pyquil/paulis.py:
```diff
@@ -119,7 +119,7 @@ def id(self, sort_ops=True):
             return ''.join("{}{}".format(p, q) for q, p in self._ops.items())
 
-    def operations_as_set(self):
+    def operations_as_set(self) -> FrozenSet[Tuple[int, str]]:
```
I think the docstring just below is wrong. It says it returns a `frozenset of (op_str, coefficient)` pairs, but your return type annotation looks correct. I think the docstring should say `frozenset of (qubit, op_str)` pairs?

rht replied:
Fixed.

rht commented:
The second commit has been squashed.

rht commented:
There is an inconsistency with `PauliTerm`'s `__init__()` where `coefficient` should have been a `Number` instead of `Union[int, float, complex]`. But I want to keep this PR compact. I will do a subsequent PR that addresses these issues later.

appleby commented:
> There is an inconsistency with `PauliTerm`'s `__init__()` where the `coefficient` argument should have been a `Number` instead of `Union[int, float, complex]`. But I want to keep this PR compact. I will do a subsequent PR that addresses these issues later.

Given [this mypy issue](https://github.com/python/mypy/issues/3186) that you pointed out in #147, which seems to imply that `numbers.Number` is somewhat useless for type checking, maybe we should go the other route and switch all the `Number`s to `Union[int, float, complex]` (as unsatisfying as that would be).

rht commented:
> Given this mypy issue that you pointed out in #147, which seems to imply that numbers.Number is somewhat useless for type checking, maybe we should go the other route and switch all the Numbers to Union[int, float, complex] (as unsatisfying as that would be).

There is an extra complication due to the fact that the code uses `isinstance(x, Number)` which works, since a `complex` object is an instance of `Number`. Using different convention for type hints (i.e. `Union[int, float, complex]`) and type checking in code would cause conflict.

rht commented:
I already have `PauliSum` annotation done. Just waiting for this PR to be merged, then I will make a new PR of `PauliSum`.

karalekas commented:
I will give this a review as soon as I've gotten @appleby's typing PR over the line -- thanks for doing this! :)

rht commented:
I just rebased to get rid of the merge conflict. @karalekas any update?

notmgsk commented:
> ... which will require you to rebase off of master first

@Tommy-Moffat enjoy!

rht commented:
@karalekas rebased and moved the line to be under `v2.14`.

karalekas commented:
@rht thanks for contributing to pyQuil! :)
